### PR TITLE
Add Assistant API for agent

### PR DESCRIPTION
**Pull request by xuechendi**
## Description

Add Assistant API for agent

## Issues

https://github.com/opea-project/GenAIComps/issues/365

## Type of change

List the type of change like below. Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds new functionality)
- [ ] Breaking change (fix or feature that would break existing design and interface)
- [ ] Others (enhancement, documentation, validation, etc.)

## Dependencies

No

## Tests

test_agent_langchain.sh

### CODE REVIEW

**Comment by xuechendi** [comment]
@minmin-intel , @XuhuiRen, @ftian1 @kevinintel @lkk12014402 
Please take a review, this PR is add Assistant API to agent

**Comment by xuechendi** [comment]
@chensuyue this UT seems to fail on other issue, may you take a look
 (Example-test / example-test (pull_request_target) Failing after 3m) seems caused by other unknown issue.

**Comment by minmin-intel** [comment]
@xuechendi I think the functionality looks good. But I have a couple of questions:
1. Did you implement the apis and memory with reference to other open-source projects or from scratch? Langgraph has support for memory and threads. See the link below. I think we should use popular open-source frameworks. It will make our agent code easier to develop and maintain. 
https://langchain-ai.github.io/langgraph/concepts/low_level/#checkpointer
https://langchain-ai.github.io/langgraph/concepts/low_level/#threads
2. Is each run saved in some kind of memory, with agent states and messages?

**Comment by xuechendi** [comment]
> @xuechendi I think the functionality looks good. But I have a couple of questions:
> 
> 1. Did you implement the apis and memory with reference to other open-source projects or from scratch? Langgraph has support for memory and threads. See the link below. I think we should use popular open-source frameworks. It will make our agent code easier to develop and maintain.
>    https://langchain-ai.github.io/langgraph/concepts/low_level/#checkpointer
>    https://langchain-ai.github.io/langgraph/concepts/low_level/#threads
> 2. Is each run saved in some kind of memory, with agent states and messages?

I have aligned with Minmin on question 1. For Question 2, run is saved in global_threads_kv which is a threadSafeDict. and it can be canceled, resume in future request.

**Comment by xuechendi** [comment]
@minmin-intel @XuhuiRen , please help to review this PR. Would like to add AssistantAPI by this week.

File: comps/agent/langchain/test.py [hunk 1]
```diff
@@ -85,6 +86,68 @@ def process_request(query):
...
+    else:
+        query = {"role": "user", "content": args.query}
+    if ret := process_request(f"threads/{thread_id}/messages", query):
+        pass
+    else:
+        print("Error when add messages !!!!")
+        return
+
+    # step 4. run
+    print("You may cancel the run with cmdline")
```
Comment (XuhuiRen) [inline comps/agent/langchain/test.py:144]: should be "running process"? Please check the gramma in the error information

**Review by XuhuiRen — APPROVED** [review]

**Review by minmin-intel — APPROVED** [review]
Looks good.

**Pull request merged**